src/utils/include/c_resource.hpp

```diff
@@ -0,0 +1,366 @@
+            }(std::forward<Ts>(Args)...);
+        }
+    }
+
+    template<typename... Ts>
+        requires(std::is_invocable_v<TConstructor, T *, Ts...>)
+    [[nodiscard]] constexpr auto emplace(Ts &&...Args) noexcept
+        requires std::is_pointer_v<T>
+    {
+        destruct(_element);
+        _element = initValue;
+        return construct(&_element, static_cast<Ts &&>(Args)...);
+    }
+
+    [[nodiscard]] constexpr c_resource(c_resource &&other) noexcept
+        requires std::is_pointer_v<T>
+    {
+        _element       = other._element;
+        other._element = initValue;
+    };
+    constexpr c_resource &operator=(c_resource &&rhs) noexcept
+        requires std::is_pointer_v<T>
+    {
+        if (this != &rhs) {
+            destruct(_element);
+            _element     = rhs._element;
+            rhs._element = initValue;
+        }
+        return *this;
+    };
+    constexpr void swap(c_resource &other) noexcept {
+        auto ptr       = _element;
+        _element       = other._element;
+        other._element = ptr;
+    }
+
+    constexpr ~c_resource() noexcept = delete;
+    constexpr ~c_resource() noexcept
+        requires destructible && std::is_pointer_v<T>
+    {
+        destruct(_element);
+    }
+    constexpr ~c_resource() noexcept
+        requires(destructible && !std::is_pointer_v<T>)
+    {
+        if constexpr (unconditionallyDestruct) {
+            destructFunction();
+        } else {
+            if (_element) {
+                destructFunction();
+            }
+        }
+    }
+
+    constexpr void clear() noexcept
+        requires destructible
+    {
+        if constexpr (std::is_pointer_v<T>) {
+            destruct(_element);
+            _element = initValue;
+        } else {
+            destructFunction();
+        }
+    }
+    constexpr c_resource &operator=(std::nullptr_t) noexcept {
+        clear();
+        return *this;
+    }
+
+    [[nodiscard]] constexpr explicit operator bool() const & noexcept {
+        return _element != initValue;
+    }
+    [[nodiscard]] constexpr explicit operator bool() && { // N.B. suppresses user-code error
+        static_assert(detail::always_false<T>, "\n**** you probably did not intend this. The resource must be assigned to an lvalue! ****\n"
+                                               "**** e.g. potential error: if (ImScoped::Window(...)) // as the temporary would be destroyed immediately ****");
+        return false;
+    }
+
+    [[nodiscard]] constexpr bool        empty() const noexcept { return _element == initValue; }
+    [[nodiscard]] constexpr friend bool have(const c_resource &r) noexcept {
+        return r._element != initValue;
+    }
+
+    auto               operator<=>(const c_resource &) = delete;
+    [[nodiscard]] bool operator==(const c_resource &rhs) const noexcept {
+        return 0 == std::memcmp(_element, rhs._element, sizeof(T));
```

Comment:
Do we need comparison? c_resource should be a unique owner of something, one instance should not be == to another. Or is there something in imgui which breaks this?

Reply:
This was in the orginal code. I guess the benefit of having comparisons is that if one has multiple references to a given resource to be able to detect that.